tools/noescape/noescape.go

```diff
@@ -0,0 +1,396 @@
+				// local escapes are the escapes found in the
+				// local function.
+				if x.Pkg.Pkg == pass.Pkg {
+					return resolve(loadFunc(x), inst.Pos())
+				}
+
+				// Recursively collect information from
+				// the other analyzers.
+				var imp packageEscapeFacts
+				if !pass.ImportPackageFact(x.Pkg.Pkg, &imp) {
+					// Is the package the sync/atomic package?
+					if x.Pkg.Pkg.Name() == "atomic" {
+						return nil // No escapes.
+					}
+					// Unable to import the dependency; we must
+					// declare these as escaping.
+					return escapes(unknownPackage, inst)
+				}
+				// The escapes of this instruction are the
+				// escapes of the called function directly.
+				return resolve(imp.Funcs[x.RelString(x.Pkg.Pkg)], inst.Pos())
+
+			case *ssa.MakeClosure:
+				// Disallow MakeClosure.
+				return escapes(closure, inst)
+			case *ssa.Builtin:
+				// Both make and new are excluded here, as they
+				// will be converted at at previous stage. We
+				// need to check for append and map creation.
+				if strings.HasPrefix(x.Name(), "append") || strings.HasPrefix(x.Name(), "makemap") {
+					return escapes(builtin, inst)
+				}
+			default:
+				// All dynamic calls are counted as soft
+				// escapes. They are similar to interface
+				// dispatches.
+				return escapes(dynamicCall, inst)
+			}
+		} else if alloc, ok := inst.(*ssa.Alloc); ok && alloc.Heap {
+			// This is a heap allocation.
+			return escapes(allocation, inst)
+		}
+		return nil // No escapes.
+	}
+
+	var analyzeBasicBlock func(*ssa.BasicBlock) []Escape // Recursive.
+	analyzeBasicBlock = func(block *ssa.BasicBlock) (rval []Escape) {
+		for _, inst := range block.Instrs {
+			rval = append(rval, analyzeInstruction(inst)...)
+		}
+		return rval // N.B. may be empty.
+	}
+
+	loadFunc = func(fn *ssa.Function) []Escape {
+		// Is this already available?
+		name := fn.RelString(pass.Pkg)
+		if es, ok := pef.Funcs[name]; ok {
+			return es
+		}
+
+		// In the case of a true cycle, we assume that the current
+		// function itself has no escapes until the rest of the
+		// analysis is complete. This will trip the above in the case
+		// of a cycle of any kind.
+		pef.Funcs[name] = nil
+
+		// Perform the basic analysis.
+		var es []Escape
+		if fn.Recover != nil {
+			es = append(es, analyzeBasicBlock(fn.Recover)...)
+		}
+		for _, block := range fn.Blocks {
+			es = append(es, analyzeBasicBlock(block)...)
+		}
+
+		// Save the result and return.
+		pef.Funcs[name] = es
+		return es
+	}
+
+	// Complete all local functions.
+	for _, fn := range state.SrcFuncs {
+		loadFunc(fn)
+	}
+
+	// Build the exception list.
+	exceptions := make(map[LinePosition]string)
+	for _, f := range pass.Files {
+		for _, cg := range f.Comments {
+			for _, c := range cg.List {
+				p := pass.Fset.Position(c.Slash)
+				if strings.HasPrefix(c.Text, "// escapes:") {
+					exceptions[LinePosition{
+						Filename: p.Filename,
+						Line:     p.Line,
+					}] = c.Text[12:] // Trim prefix.
```

Comment:
Optional: use `len("// +noescape")` for clarity.

Reply:
Done.